filc/retrieval.go

```diff
@@ -0,0 +1,416 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/application-research/filclient"
+	"github.com/application-research/filclient/retrievehelper"
+	"github.com/dustin/go-humanize"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-fil-markets/retrievalmarket"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ipldformat "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	"github.com/ipld/go-ipld-prime"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	"golang.org/x/xerrors"
+)
+
+type RetrievalCandidate struct {
+	Miner   address.Address
+	RootCid cid.Cid
+	DealID  uint
+}
+
+type CandidateSelectionConfig struct {
+	// Whether retrieval over IPFS is preferred if available
+	tryIPFS bool
+
+	// If true, candidates will be tried in the order they're passed in
+	// unchanged (and all other sorting-related options will be ignored)
+	noSort bool
+}
+
+type RetrievalResults struct {
+}
+
+func (node *Node) GetRetrievalCandidates(endpoint string, c cid.Cid) ([]RetrievalCandidate, error) {
+
+	endpointURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, xerrors.Errorf("endpoint %s is not a valid url", endpoint)
+	}
+	endpointURL.Path = path.Join(endpointURL.Path, c.String())
+
+	resp, err := http.Get(endpointURL.String())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("http request to endpoint %s got status %v", endpointURL, resp.StatusCode)
+	}
+
+	var res []RetrievalCandidate
+
+	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return nil, xerrors.Errorf("could not unmarshal http response for cid %s", c)
+	}
+
+	return res, nil
+}
+
+type RetrievalStats interface {
+	GetByteSize() uint64
+	GetDuration() time.Duration
+	GetAverageBytesPerSecond() uint64
+}
+
+type FILRetrievalStats struct {
+	filclient.RetrievalStats
+}
+
+func (stats *FILRetrievalStats) GetByteSize() uint64 {
+	return stats.Size
+}
+
+func (stats *FILRetrievalStats) GetDuration() time.Duration {
+	return stats.Duration
+}
+
+func (stats *FILRetrievalStats) GetAverageBytesPerSecond() uint64 {
+	return stats.AverageSpeed
+}
+
+type IPFSRetrievalStats struct {
+	ByteSize uint64
+	Duration time.Duration
+}
+
+func (stats *IPFSRetrievalStats) GetByteSize() uint64 {
+	return stats.ByteSize
+}
+
+func (stats *IPFSRetrievalStats) GetDuration() time.Duration {
+	return stats.Duration
+}
+
+func (stats *IPFSRetrievalStats) GetAverageBytesPerSecond() uint64 {
+	return uint64(float64(stats.ByteSize) / stats.Duration.Seconds())
+}
+
+// NOTE (6 Oct 2021) - presently, if selNode is specified, then cfg.tryIPFS will
+// be silently ignored, and an ipfs retrieval will not be attempted. This will
```

Comment:
See my suggestion above: never "silently ignore", instead have your distant functions in a position to always receive reasonable inputs ( and double down on checking + erroring when given nonsensical inputs )

Fix the comment once you fix the cli parsing